File: bun-version/src/providers/tencentmaas.ts
import { BaseSearchProvider } from "./base";
import type { WebSearchResult } from "../types";

export class TencentmaasProvider extends BaseSearchProvider {
  readonly name = "tencentmaas";
  readonly toolName = "tencentmaas-search";
  readonly description = "腾讯云联网搜索 (混元)";
  readonly requiredParams = ["tencentmaas-apiKey"];
  readonly optionalParams = [
    { name: "model", description: "模型名称", defaultValue: "hy3" },
    {
      name: "search_source",
      description: "搜索版本 (lite/standard)",
      defaultValue: "standard",
    },
  ];

  async search(
    query: string,
    urlParams: URLSearchParams
  ): Promise<WebSearchResult> {
    const apiKey = urlParams.get("tencentmaas-apiKey")!;
    const model = this.getParam(urlParams, "model", "hy3")!;
    const searchSource = this.getParam(
      urlParams,
      "search_source",
      "standard"
    )!;

    console.log(`[tencentmaas-search] query: "${query}", model: ${model}`);

    const resp = await fetch(
      "https://tokenhub.tencentmaas.com/v1/chat/completions",
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          model,
          messages: [
            { role: "system", content: "You are a helpful assistant." },
            { role: "user", content: query },
          ],
          web_search_options: {
            enable: true,
            search_source: searchSource,
          },
          stream: false,
        }),
      }
    );

    if (!resp.ok) {
      throw new Error(
        `tencentmaas-search API error: ${resp.status} ${await resp.text()}`
      );
    }

    const data = (await resp.json()) as any;
    const text = data.choices?.[0]?.message?.content ?? "";

    return {
      content: [{ type: "text", text: text || "未获取到搜索结果" }],
    };
  }
}

File: src/provider/tencentmaas.rs
use super::{SearchProvider, WebSearchResult};
use serde_json::json;
use std::collections::HashMap;

pub struct TencentmaasProvider;

#[async_trait::async_trait]
impl SearchProvider for TencentmaasProvider {
    fn name(&self) -> &str {
        "tencentmaas"
    }

    fn tool_name(&self) -> &str {
        "tencentmaas-search"
    }

    fn description(&self) -> &str {
        "腾讯云联网搜索 (混元)"
    }

    fn required_params(&self) -> &[&str] {
        &["tencentmaas-apiKey"]
    }

    async fn search(
        &self,
        query: &str,
        params: &HashMap<String, String>,
        client: &reqwest::Client,
    ) -> Result<WebSearchResult, String> {
        let api_key = params
            .get("tencentmaas-apiKey")
            .ok_or_else(|| "Missing tencentmaas-apiKey".to_string())?;

        let model = params
            .get("tencentmaas-model")
            .map(|s| s.as_str())
            .unwrap_or("hy3");

        let search_source = params
            .get("tencentmaas-search_source")
            .map(|s| s.as_str())
            .unwrap_or("standard");

        println!(
            "[tencentmaas-search] query: \"{}\", model: {}",
            query, model
        );

        let body = json!({
            "model": model,
            "messages": [
                { "role": "system", "content": "You are a helpful assistant." },
                { "role": "user", "content": query }
            ],
            "web_search_options": {
                "enable": true,
                "search_source": search_source
            },
            "stream": false
        });

        let res = client
            .post("https://tokenhub.tencentmaas.com/v1/chat/completions")
            .bearer_auth(api_key)
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("Network error: {}", e))?;

        if !res.status().is_success() {
            let err_text = res.text().await.unwrap_or_default();
            return Err(format!("tencentmaas-search API error: {}", err_text));
        }

        let data: serde_json::Value = res
            .json()
            .await
            .map_err(|e| format!("Failed to parse response: {}", e))?;

        let text = data["choices"]
            .get(0)
            .and_then(|c| c["message"]["content"].as_str())
            .unwrap_or("未获取到搜索结果");

        Ok(WebSearchResult::text(text))
    }
}
